src/WinRT.Runtime2/InteropServices/RestrictedErrorInfo.cs

```diff
@@ -62,26 +294,155 @@ public static void ThrowExceptionForHR(HRESULT errorCode)
     /// <seealso cref="Marshal.GetExceptionForHR(int)"/>
     public static HRESULT GetHRForException(Exception? exception)
     {
-        // TODO
-        return 0;
+        if (exception == null)
+        {
+            throw new ArgumentNullException("exception");
+        }
+        int hr = exception.HResult;
+        try
+        {
+            if (exception.TryGetRestrictedLanguageErrorInfo(out WindowsRuntimeObjectReference? restrictedErrorObject, out bool _))
+            {
+                if (restrictedErrorObject != null)
+                {
+                    IRestrictedErrorInfoMethods.GetErrorDetails(restrictedErrorObject.GetThisPtr(), out hr);
+                    GC.KeepAlive(restrictedErrorObject);
+                }
+            }
+        }
+        catch (Exception e)
+        {
+            // If we fail to get the hresult from the error info, we fallback to the exception hresult.
+            Debug.Assert(false, e.Message, e.StackTrace);
+        }
+
+        return hr switch
+        {
+            WellKnownErrorCodes.COR_E_OBJECTDISPOSED => WellKnownErrorCodes.RO_E_CLOSED,
+            WellKnownErrorCodes.COR_E_OPERATIONCANCELED => WellKnownErrorCodes.ERROR_CANCELLED,
+            WellKnownErrorCodes.COR_E_ARGUMENTOUTOFRANGE or WellKnownErrorCodes.COR_E_INDEXOUTOFRANGE => WellKnownErrorCodes.E_BOUNDS,
+            WellKnownErrorCodes.COR_E_TIMEOUT => WellKnownErrorCodes.ERROR_TIMEOUT,
+            _ => hr,
+        };
     }
 
     /// <summary>
     /// Stores info on the input exception through the <c>IRestrictedErrorInfo</c> infrastructure, to retrieve it later.
     /// </summary>
     /// <param name="exception">The input <see cref="Exception"/> instance to store.</param>
-    public static void SetErrorInfo(Exception exception)
+    public static unsafe void SetErrorInfo(Exception exception)
     {
-        // TODO
+        try
+        {
+            if (WindowsRuntimeImports.GetRestrictedErrorInfo != null && WindowsRuntimeImports.SetRestrictedErrorInfo != null && WindowsRuntimeImports.RoOriginateLanguageException != null)
```

Comment:
I suspect the implicit delegate conversion is tripping you up here, as these checks aren't valid (as in, these are now methods, they can never be null). Basically you can delete any checks like these that we had from before 🙂